### PR TITLE
Allow multiple LDAP user bind patterns in config

### PR DESCRIPTION
**Pull request by jbeluch**
Hi, at our company we have a need to try multiple bind patterns as we have accounts with different DN hierarchies. Is this a feature you would be interested in taking? Thanks!

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapAuthenticator.java [hunk 1]
```diff
@@ -188,7 +190,32 @@ private static String decodeCredentials(String base64EncodedCredentials)
     private Principal authenticate(String user, String password)
             throws AuthenticationException
     {
-        Map<String, String> environment = createEnvironment(user, password);
+        AuthenticationException ex = null;
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/security/LdapAuthenticator.java]: As I understand, and based on reading [this](https://docs.oracle.com/javase/tutorial/jndi/ldap/exceptions.html), there are two types of LDAP errors:

* `javax.naming.AuthenticationException`: Invalid credentials. Prompt the user again for credentials.
* Any other `NamingException`: There was an error talking to LDAP, or the LDAP server doesn't like us, etc., so having valid credentials (which we might have) won't help, no need to prompt the user.

We should consider how that should be handled here. If the first call fails due to invalid credentials, and the second one fails due to another error (maybe transient), should we prompt the user? Also, what error should we return? All of them? The last one? Combine together with de-duplication? Or maybe we shouldn't return the error from the LDAP server (not sure that is a good idea anyway).

Note that we prompt the user by setting an HTTP authenticate header value when throwing `AuthenticationException`. For the Presto CLI and JDBC driver, we always pre-authenticate, so prompting is ignored. This is recommend for performance reasons (avoids a round-trip), and it's possible most/all clients do this, so the prompting part might not matter.

This is more relevant with #7632 which splits out LDAP with an API that's decoupled from HTTP.

File: presto-main/etc/log.properties [hunk 2]
```diff
@@ -9,3 +9,4 @@ com.facebook.presto=INFO
 com.sun.jersey.guice.spi.container.GuiceComponentProviderFactory=WARN
 com.ning.http.client=WARN
 com.facebook.presto.server.PluginManager=DEBUG
+com.facebook.presto.server.security.LdapAuthenticator=DEBUG
```
Comment (electrum) [inline presto-main/etc/log.properties]: Let's remove this, as it's not broadly useful for development. We expect people to add things as needed while developing, without having everything enabled by default.

**Comment by electrum** [comment]
Please also add an example of multiple patterns in the LDAP documentation in `presto-docs/src/main/sphinx/security/ldap.rst`

File: presto-main/src/main/java/com/facebook/presto/server/security/LdapAuthenticator.java [hunk 3]
```diff
@@ -198,31 +225,31 @@ private Principal authenticate(String user, String password)
             return new LdapPrincipal(user);
         }
         catch (javax.naming.AuthenticationException e) {
-            log.debug("Authentication failed for user [%s]: %s", user, e.getMessage());
+            log.debug("Authentication failed for user [%s]: %s", environment.get(SECURITY_PRINCIPAL), e.getMessage());
```
[the remark this answers is not in the document]
Comment (anusudarsan) [reply presto-main/src/main/java/com/facebook/presto/server/security/LdapAuthenticator.java]: Also why do you want to log the entire principal? It looks less secure to have the log print out the entire domain/dn of the ldap server.

[the remark this answers is not in the document]
Comment (jbeluch) [reply presto-main/src/main/java/com/facebook/presto/server/security/LdapAuthenticator.java]: It helps with debugging to know which ldap bind pattern failed. It seems to me that distinguished names do not need to be treated as a secret.

**Comment by jbeluch** [comment]
Thanks for the feedback @electrum, will work on the suggestions.

**Comment by electrum** [comment]
I had some time on a recent flight to split out a password authenticator plugin #9865. It might be easier to rebase this PR on top of that change, since it cleans up the LDAP code a bit and makes the authentication interface clearer.

**Comment by rentaow** [comment]
Can we revisit this PR? We also have a need for this feature.

**Comment by jbeluch** [comment]
@rentaow I've been meaning to find some time... Hopefully get to it next week if no else does first.

**Pull request closed without merging**